game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java

```diff
@@ -43,88 +43,85 @@
-  public static final ClientSetting testLobbyHost = new ClientSetting("TEST_LOBBY_HOST");
-  public static final ClientSetting testLobbyPort = new ClientSetting("TEST_LOBBY_PORT");
-  public static final ClientSetting firstTimeThisVersion =
-      new ClientSetting("TRIPLEA_FIRST_TIME_THIS_VERSION_PROPERTY", true);
-  public static final ClientSetting lastCheckForEngineUpdate =
-      new ClientSetting("TRIPLEA_LAST_CHECK_FOR_ENGINE_UPDATE");
-  public static final ClientSetting lastCheckForMapUpdates = new ClientSetting("TRIPLEA_LAST_CHECK_FOR_MAP_UPDATES");
-  public static final ClientSetting promptToDownloadTutorialMap =
-      new ClientSetting("TRIPLEA_PROMPT_TO_DOWNLOAD_TUTORIAL_MAP", true);
-  public static final ClientSetting userMapsFolderPath = new ClientSetting(
+public abstract class ClientSetting<T> implements GameSetting<T> {
+  public static final ClientSetting<String> aiPauseDuration = new StringClientSetting("AI_PAUSE_DURATION", 400);
+  public static final ClientSetting<String> arrowKeyScrollSpeed = new StringClientSetting("ARROW_KEY_SCROLL_SPEED", 70);
+  public static final ClientSetting<String> battleCalcSimulationCountDice =
+      new StringClientSetting("BATTLE_CALC_SIMULATION_COUNT_DICE", 200);
+  public static final ClientSetting<String> battleCalcSimulationCountLowLuck =
+      new StringClientSetting("BATTLE_CALC_SIMULATION_COUNT_LOW_LUCK", 500);
+  public static final ClientSetting<String> confirmDefensiveRolls =
+      new StringClientSetting("CONFIRM_DEFENSIVE_ROLLS", false);
+  public static final ClientSetting<String> confirmEnemyCasualties =
+      new StringClientSetting("CONFIRM_ENEMY_CASUALTIES", false);
+  public static final ClientSetting<String> defaultGameName =
+      new StringClientSetting("DEFAULT_GAME_NAME_PREF", "Big World : 1942");
+  public static final ClientSetting<String> defaultGameUri = new StringClientSetting("DEFAULT_GAME_URI_PREF");
+  public static final ClientSetting<String> fasterArrowKeyScrollMultiplier =
+      new StringClientSetting("FASTER_ARROW_KEY_SCROLL_MULTIPLIER", 2);
+  public static final ClientSetting<String> spaceBarConfirmsCasualties =
+      new StringClientSetting("SPACE_BAR_CONFIRMS_CASUALTIES", true);
+  public static final ClientSetting<String> lobbyLastUsedHost = new StringClientSetting("LOBBY_LAST_USED_HOST");
+  public static final ClientSetting<String> lobbyLastUsedPort = new StringClientSetting("LOBBY_LAST_USED_PORT");
+  public static final ClientSetting<String> lookAndFeel =
+      new StringClientSetting("LOOK_AND_FEEL_PREF", LookAndFeel.getDefaultLookAndFeelClassName());
+  public static final ClientSetting<String> mapEdgeScrollSpeed = new StringClientSetting("MAP_EDGE_SCROLL_SPEED", 30);
+  public static final ClientSetting<String> mapEdgeScrollZoneSize =
+      new StringClientSetting("MAP_EDGE_SCROLL_ZONE_SIZE", 30);
+  public static final ClientSetting<String> mapFolderOverride = new StringClientSetting("MAP_FOLDER_OVERRIDE");
```

Comment:
> There seems to be a lot of opportunity for stronger typing. For example the folder override could be a File, the L&F preference could be a look and feel object, the scroller speeds could be integers, etc...

Agreed.  That's what I proposed to do above in a series of follow-up PRs.

> Will creating a child class for each type scale well?
>
> It looks like the thing that is different is how a value is extracted. What are your thoughts to parameterize the value extraction @ssoloff instead of using child instances?

I currently only see specializations for the following types:

* `String`
* `Integer`
* `Boolean`
* `File`
* `Enum` (I believe all enums can be handled uniformly)

Instead of subclasses implementing template methods defined by the superclass, we could parameterize the conversion behavior using the _Strategy_ pattern by passing a `Function<T, String>` and a  `Function<String, T>` to the `ClientSetting` constructor, as you suggested.

However, because all `ClientSetting`s of the same generic type `T` will almost assuredly use the same behavior, we wouldn't want to have to specify these functions for every `ClientSetting` instance we create.  Instead, we could define factory methods for each type that encapsulate the conversion functions .  For example:

```java
public static final ClientSetting<Integer> mapEdgeScrollSpeed = newIntegerClientSetting("MAP_EDGE_SCROLL_SPEED", 30);
public static final ClientSetting<String> mapFolderOverride = newStringClientSetting("MAP_FOLDER_OVERRIDE");
```

I'm not sure what's the best solution at this time.  I would suggest to proceed with the current proposal for now, and once more settings are converted to their type-safe equivalents, I will re-evaluate if behavior parameterization via _Strategy_ is a better choice.